### PR TITLE
Remote creation API

### PR DESCRIPTION
**Pull request by tiennou**
This is the PR for the code I've written for #4658.

It adds a `git_remote_create_ext` with an option structure for handling the 3 other `_create` calls (ie. `_with_fetchspec`, `_anonymous`, `_detached`).

I've not pondered too much about the difference between anonymous & detached remotes. AFAICT, both are in-memory, and anonymous ones can make use of an underlying repository, so config will be respected.

### CODE REVIEW

**Comment by vors** [comment]
😍 thank you for taking a stub at it!

File: src/remote.c [hunk 1]
```diff
@@ -189,6 +189,28 @@ static int canonicalize_url(git_buf *out, const char *in)
...
+		return 0;
+
+	if (error < 0)
+		return error;
+
+	git_remote_free(remote);
+
+	giterr_set(
+			   GITERR_CONFIG,
+			   "remote '%s' already exists", name);
```
Comment (pks-t) [inline src/remote.c]: Indentation here is a bit funny.

File: include/git2/remote.h [hunk 2]
```diff
@@ -41,6 +41,52 @@ GIT_EXTERN(int) git_remote_create(
...
+	const char *name;
+
+	/** The fetchspec the remote should use. */
+	const char *fetchspec;
+} git_remote_create_options;
+
+#define GIT_REMOTE_CREATE_OPTIONS_VERSION 1
+#define GIT_REMOTE_CREATE_OPTIONS_INIT {GIT_REMOTE_CREATE_OPTIONS_VERSION}
+
+GIT_EXTERN(int) git_remote_create_init_options(
```
Comment (pks-t) [inline include/git2/remote.h]: This function is missing its documentation.

File: include/git2/remote.h [hunk 3]
```diff
@@ -41,6 +41,52 @@ GIT_EXTERN(int) git_remote_create(
...
+ * Remote creation options structure
+ *
+ * Initialize with `GIT_REMOTE_CREATE_OPTIONS_INIT`. Alternatively, you can
+ * use `git_remote_create_init_options`.
+ *
+ */
+typedef struct git_remote_create_options {
+	unsigned int version;
+
+	/** The repository that should own the remote.
```
Comment (pks-t) [inline include/git2/remote.h]: Our struct member doc comment style is to either have a single `/** foobar */` line or start with an empty `/**` line.

File: include/git2/remote.h [hunk 4]
```diff
@@ -41,6 +41,52 @@ GIT_EXTERN(int) git_remote_create(
...
+ * Create a remote, with options.
+ *
+ * This function allows more fine-grained control over the remote creation.
+ *
+ * @param out the resulting remote
+ * @param url the remote's url
+ * @param opts the remote creation options
+ * @return 0 on success, WIP
+ */
+GIT_EXTERN(int) git_remote_create_ext(
```
Comment (pks-t) [inline include/git2/remote.h]: Bikeshedding: `create_ext` doesn't make it clear what the "extended" nature of this function is. Instead, I'd vote for `git_remote_create_with_opts`, which clearly states that it takes an `opts` structure.

File: include/git2/remote.h [hunk 5]
```diff
@@ -41,6 +41,52 @@ GIT_EXTERN(int) git_remote_create(
...
+		unsigned int version);
+
+/**
+ * Create a remote, with options.
+ *
+ * This function allows more fine-grained control over the remote creation.
+ *
+ * @param out the resulting remote
+ * @param url the remote's url
+ * @param opts the remote creation options
```
Comment (pks-t) [inline include/git2/remote.h]: You should mention that `opts` is optional and what will happen in that case. (It is optional, isn't it? If not, I'd propose to just create an anonymous remote in that case)

File: src/remote.c [hunk 6]
```diff
@@ -211,58 +211,79 @@ static int ensure_remote_doesnot_exist(git_repository *repo, const char *name)
...
 			goto on_error;
 
 		/* only write for non-anonymous remotes */
-		if (repo && name && (error = write_add_refspec(repo, name, fetch, true)) < 0)
-			goto on_error;
-
-		if (repo && (error = lookup_remote_prune_config(remote, config_ro, name)) < 0)
+		if (!is_anonymous &&
+			((error = write_add_refspec(opts->repository, opts->name, opts->fetchspec, true)) < 0 ||
+			(error = lookup_remote_prune_config(remote, config_ro, opts->name)) < 0))
```
Comment (pks-t) [inline src/remote.c]: Wait, this is different from what we did before. Now we only call `lookup_remote_prune_config` if both `opts->repository` and `opts->name` are set. Previously, we were also calling it in case `opts->name` is unset. Is this intentional and, if so, why can we avoid calling `lookup_remote_prune_config` now?

File: src/remote.c [hunk 7]
```diff
@@ -271,7 +292,7 @@ static int create_internal(git_remote **out, git_repository *repo, const char *n
 	}
 
 	/* A remote without a name doesn't download tags */
-	if (!name)
+	if (is_anonymous)
```
Comment (pks-t) [inline src/remote.c]: This is also different from what we did before, as now an unnamed remote with an associated repository will not fetch tags anymore. Is this intended?

File: src/remote.c [hunk 8]
```diff
@@ -306,37 +327,41 @@ int git_remote_create(git_remote **out, git_repository *repo, const char *name,
...
 
-	if (create_internal(&remote, repo, name, url, fetch) < 0)
-		goto on_error;
-
-	*out = remote;
-
-	return 0;
+	opts.repository = repo;
+	opts.name = name;
+	opts.fetchspec = fetch;
```
Comment (pks-t) [inline src/remote.c]: This conversion is great

File: src/remote.c [hunk 9]
```diff
@@ -306,37 +327,41 @@ int git_remote_create(git_remote **out, git_repository *repo, const char *name,
...
 {
-	git_remote *remote = NULL;
 	int error;
+	git_remote_create_options opts = GIT_REMOTE_CREATE_OPTIONS_INIT;
 
 	if ((error = ensure_remote_name_is_valid(name)) < 0)
 		return error;
 
 	if ((error = ensure_remote_doesnot_exist(repo, name)) < 0)
 		return error;
```
Comment (pks-t) [inline src/remote.c]: These two calls are already done in `create_remote_internal` now. I'd like to avoid this duplication

File: include/git2/remote.h [hunk 10]
```diff
@@ -63,6 +73,9 @@ typedef struct git_remote_create_options {
 
 	/** The fetchspec the remote should use. */
 	const char *fetchspec;
+
+	/** Additional flags for the remote. See git_remote_create_flags. */
+	int flags;
```
Comment (pks-t) [inline include/git2/remote.h]: This should be unsigned, shouldn't it?

File: src/remote.c (hunk 6 above)
Comment (tiennou) [reply src/remote.c]: Yeah, the tests around detached/anonymous/realized remotes are confusing here. My `is_anonymous` variable should be named `is_not_realized` maybe (bikeshedding welcome on the name).

The original code would attempt to load a `remote..prune` config option, which IMHO doesn't make sense (and I remember a test broke before I changed it to this).

File: src/remote.c (hunk 7 above)
Comment (tiennou) [reply src/remote.c]: You're right, I went too far trying to make sense of the 3 types of remotes. Reverting.

File: src/remote.c [hunk 11]
```diff
@@ -211,58 +211,79 @@ static int ensure_remote_doesnot_exist(git_repository *repo, const char *name)
...
+		(error = git_repository_config_snapshot(&config_ro, opts->repository)) < 0)
 		return error;
 
 	remote = git__calloc(1, sizeof(git_remote));
 	GITERR_CHECK_ALLOC(remote);
 
-	remote->repo = repo;
+	remote->repo = opts->repository;
 
 	if ((error = git_vector_init(&remote->refs, 32, NULL)) < 0 ||
```
Comment (tiennou) [inline src/remote.c]: Just a heads up, I was pretty surprised about that 32 here. Should I lower that to something like 4 ?

File: src/remote.c (hunk 6 above)
Comment (tiennou) [reply src/remote.c]: Confirmed. Should I keep the test as is ?
```
  1) Failure:
network::remote::remotes::cannot_add_a_nameless_remote [/Users/tiennou/Projects/libgit2/tests/network/remote/remotes.c:321]
  GIT_EINVALIDSPEC != git_remote_create(&remote, _repo, NULL, "git://github.com/libgit2/libgit2")
  -12 != 0
```

**Comment by tiennou** [comment]
Rebased, with all review comments taken care of apart from the change of behavior [here](https://github.com/libgit2/libgit2/pull/4667/files#diff-2f07edcf8df8f205905e1e7d24631753R284).

File: src/remote.c (hunk 6 above)
Comment (pks-t) [reply src/remote.c]: You're right, the original code doesn't make a lot of sense. So this change looks correct to me

File: tests/network/remote/remotes.c [hunk 12]
```diff
@@ -318,7 +318,7 @@ void test_network_remote_remotes__cannot_add_a_nameless_remote(void)
 
 	cl_assert_equal_i(
 		GIT_EINVALIDSPEC,
-		git_remote_create(&remote, _repo, NULL, "git://github.com/libgit2/libgit2"));
+		git_remote_create(&remote, _repo, "", "git://github.com/libgit2/libgit2"));
```
Comment (pks-t) [inline tests/network/remote/remotes.c]: This is the test you've been talking about, right? I'd say we should just keep this test as it was originally, just to be sure we're not altering behaviour.

File: include/git2/remote.h [hunk 13]
```diff
@@ -41,6 +41,16 @@ GIT_EXTERN(int) git_remote_create(
 		const char *name,
 		const char *url);
 
+/**
+ * Remote creation options flags
+ */
+typedef enum {
+	GIT_REMOTE_CREATE_DEFAULT = 0,
```
Comment (pks-t) [inline include/git2/remote.h]: Why is this flag necessary? We don't have default flags for all the other option enums, do we?

File: src/remote.c (hunk 6 above)
Comment (tiennou) [reply src/remote.c]: Note that most of what I said above about the broken test is wrong. The test failure was caused when I removed the name & repo existence checks from `git_remote_create` (since it allows `NULL`).

AFAICT, the original code is extra-confusing about what actually needs a repo/name because it indirectly uses the config, so I hope this is better. I'm still dubious of the whole config_ro/config_rw dance that's happening, because we start by doing a snapshot, then we re-grab a "live" config. I might be just paranoid/unaware of the semantics of `config__weakptr`, but I'm tempted to change it to : grab a "live" version IF we have a repo, snapshot it so it doesn't change when applying `url.insteadof`, and set the various other settings on the first live one.

**Comment by tiennou** [comment]
So, major rework of this. I've taken the time to strengthen the test suite because I kept discovering "edge-cases" (like with_fetchspec not building a default fetchspec), so all 3 basic calls are now more cautiously tested (for invalid parameters, config changes, and whatever differing behavior I stumbled upon).

New from last time :
- I've added a flag to prevent building a default fetchspec. I've also re-duplicated code either for backward-compatibility reasons (`git_remote_create` actually shadows its error codes) or simplicity in `create_internal`.
- I've switched the config-management code to use transactions (so the setup/teardown cycle is simpler).

Left to do/open questions :
- check that we set error messages correctly. I noticed a few `<no message>` failures.
- maybe have the `const char *fetchspec` option be a `size_t count, const char fetchspecs[]` ? I didn't notice at first, but using `git_remote_add_fetch` will need a relookup so the refspecs are "in the correct place" (see `dwim_refspecs` at the end of `create_internal`).

File: tests/remote/create.c [hunk 14]
```diff
@@ -27,11 +37,68 @@ void test_remote_create__manual(void)
...
+	cl_git_pass(git_config_get_string(&cfg_val, cfg, "remote.valid-name.url"));
+	cl_assert_equal_s(cfg_val, TEST_URL);
+
+	git_config_free(cfg);
+	git_remote_free(remote);
+}
+
+void test_remote_create__named_fail_on_invalid_name(void)
+{
+	cl_git_assert_cannot_create_remote(GIT_EINVALIDSPEC, git_remote_create(&r, _repo, NULL, TEST_URL));
```
Comment (pks-t) [inline tests/remote/create.c]: I'd actually prefer the same as in the other PR. Instead of using a macro and repeating that multiple times, you should just use an array of strings here and iterate through them. That'd be a lot simpler to read and produce less code.

File: src/remote.c [hunk 15]
```diff
@@ -232,7 +232,7 @@ static int create_internal(git_remote **out, git_repository *repo, const char *n
 
 	remote->repo = repo;
 
-	if ((error = git_vector_init(&remote->refs, 32, NULL)) < 0 ||
+	if ((error = git_vector_init(&remote->refs, 8, NULL)) < 0 ||
```
Comment (pks-t) [inline src/remote.c]: I'd really like to see some reasoning in the commit message as to why this is a good thing.

File: src/remote.c [hunk 16]
```diff
@@ -278,19 +281,19 @@ static int create_internal(git_remote **out, git_repository *repo, const char *n
...
+		git_config_unlock(config, 1);
 	*out = remote;
 	error = 0;
 
 on_error:
 	if (error)
 		git_remote_free(remote);
 
-	git_config_free(config_ro);
+	git_config_free(config);
```
Comment (pks-t) [inline src/remote.c]: Don't we have to unlock the config on error, too?

File: src/remote.c (hunk 15 above)
Comment (tiennou) [reply src/remote.c]: Memory bloat ? That's enough memory for 32 different refspecs. I expect this is far larger than any "normal" usage of a remote. 8 is selected because it's the minimum allocation size of `git_vector`.

File: src/remote.c (hunk 16 above)
Comment (tiennou) [reply src/remote.c]: Will fix, I was misusing the transaction config API (hence the CI failure). The transaction handles unlocking when it's freed.

File: tests/remote/create.c (hunk 14 above)
Comment (tiennou) [reply tests/remote/create.c]: Fixed.

File: include/git2/remote.h (hunk 13 above)
Comment (tiennou) [reply include/git2/remote.h]: Fixed.

File: src/remote.c (hunk 15 above)
Comment (pks-t) [reply src/remote.c]: The minimum allocation size is the interesting information I've been looking for :) I wasn't really wondering why the count was removed, but rather why 8 was picked. Thanks for clarifying

File: src/remote.c [hunk 17]
```diff
@@ -299,47 +321,60 @@ int git_remote_create(git_remote **out, git_repository *repo, const char *name,
...
+	git_remote_create_options opts = GIT_REMOTE_CREATE_OPTIONS_INIT;
 
-	if (default_fetchspec_for_name(&buf, name) < 0)
-		return -1;
+	/* Those 2 tests are duplicated here because of backward-compatibility */
+	if ((error = ensure_remote_name_is_valid(name)) < 0)
+		return error;
+
+	if (canonicalize_url(&buf, url) < 0)
+		return GIT_ERROR;
```
Comment (pks-t) [inline src/remote.c]: What kind of backwards compatibility do you refer to? Do callers expect a specific error code here?

File: src/remote.c [hunk 18]
```diff
@@ -263,7 +263,7 @@ static int create_internal(git_remote **out, const char *url, const git_remote_c
 		(error = canonicalize_url(&canonical_url, url)) < 0)
 		goto on_error;
 
-	if (config) {
+	if (config && (opts->flags & GIT_REMOTE_CREATE_SKIP_INSTEADOF) != GIT_REMOTE_CREATE_SKIP_INSTEADOF) {
```
Comment (pks-t) [inline src/remote.c]: `if (config && !(opts->flags & GIT_REMOTE_CREATE_SKIP_INSTEADOF))`

File: src/remote.c [hunk 19]
```diff
@@ -280,13 +281,24 @@ static int create_internal(git_remote **out, const char *url, const git_remote_c
 			goto on_error;
 	}
 
-	if (opts->fetchspec != NULL) {
-		if ((error = add_refspec(remote, opts->fetchspec, true)) < 0)
+	if (opts->fetchspec != NULL ||
+	    (opts->name && (opts->flags & GIT_REMOTE_CREATE_SKIP_DEFAULT_FETCHSPEC) != GIT_REMOTE_CREATE_SKIP_DEFAULT_FETCHSPEC)) {
```
Comment (pks-t) [inline src/remote.c]: `(opts->name && !(opts->flags & GIT_REMOTE_CREATE_SKIP_DEFAULT_FETCHSPEC))`

File: src/remote.c (hunk 17 above)
Comment (tiennou) [reply src/remote.c]: I found out while writing [this test](https://github.com/libgit2/libgit2/pull/4667/commits/eb725118f18a92cfcf05de58f4ff90139af9c82f#diff-0f6c0f30ddbd94614e1f4f4887955556R93) that [this](https://github.com/libgit2/libgit2/blob/master/src/remote.c#L322) was "misbehaving", in that it never returns what `create_internal` says when it fails. It depends on what kind of backward compatibility we want (I realize I may be too zealous, but maybe this should be discussed separately)…

The `ensure_remote_name_is_valid` is duplicated in case name is NULL (a NULL name in the `with_opts` version would cause an anonymous remote to be created instead, ie. it's not a failure anymore). An alternative would be to assert on the mandatory parameters to be correct…

**Comment by tiennou** [comment]
Should be good now. I've lowered the vector size again, because (if my sizeof math is correct), we have 32 * sizeof(void *) => 256bytes on 64bits, which makes 1Mb lost to mostly empty vectors for every 4 remote you create.

Thanks @pks-t for the reviews and ever helpful comments.

File: src/remote.c [hunk 20]
```diff
@@ -278,19 +280,19 @@ static int create_internal(git_remote **out, git_repository *repo, const char *n
 	else
 		remote->download_tags = GIT_REMOTE_DOWNLOAD_TAGS_AUTO;
 
-
-	git_buf_dispose(&var);
-
+	if (txn)
+		git_transaction_commit(txn);
```
Comment (pks-t) [inline src/remote.c]: Sorry to not have catched this earlier, but you need to do error checking here. Otherwise we might end up silently not writing the config changes.

File: src/remote.c (hunk 17 above)
Comment (pks-t) [reply src/remote.c]: Yeah. I guess we should fix this, but that can be done in a later PR if you wish.

**Comment by tiennou** [comment]
Just a thought : per https://stackoverflow.com/a/51570151/444323, if I kept the new function's prototype compatible with clone's `git_remote_create_cb`, that would make a nice one-shot customization function, since `git_remote_create_opts` could be passed as `payload`.

**Comment by tiennou** [comment]
@pks-t If you don't mind (and your workqueue is not saturated), I'd like to have this one reviewed 😉.

File: src/remote.c [hunk 21]
```diff
@@ -189,58 +185,121 @@ static int canonicalize_url(git_buf *out, const char *in)
...
+	    (opts->name && !(opts->flags & GIT_REMOTE_CREATE_SKIP_DEFAULT_FETCHSPEC))) {
+		const char *fetch = NULL;
+		if (opts->fetchspec) {
+			fetch = opts->fetchspec;
+		} else {
+			if ((error = default_fetchspec_for_name(&specbuf, opts->name)) < 0)
+				goto on_error;
+
+			fetch = git_buf_cstr(&specbuf);
+		}
```
[the remark this answers is not in the document]
Comment (tiennou) [reply src/remote.c]: I don't expect it does : the `on_error` "handler" unconditionally frees the buffer, so it doesn't matter if it's a self-built fetchspec or not. Arguably, I could rename that to `cleanup`…

**Comment by tiennou** [comment]
Okay, something reaally weird is going on after rebasing to `master`… Just to be sure, I'll submit the rebased version as another "dummy" PR, and I'm restoring this one to the pre-rebase version (though the review comments have been taken care of) because I'd like to make sure it's still green (IIRC it was, but 🤷‍♂️ ). Please don't merge this while investigations are in progress…

**Comment by tiennou** [comment]
Heh, since VSTS does a merge to master beforehand (obviously), it doesn't make sense to dummy-PR anymore (the branch is still up [here](https://github.com/libgit2/libgit2/compare/master...tiennou:feature/remote-create-api-master). I've managed to confused the build view so here's a [link](https://dev.azure.com/libgit2/libgit2/_build/results?buildId=741&view=logs) to its last results.

My initial reaction was to look at the leaks (since @pks-t was actually worried about those fetchspecs leaking and was thus quite surprised), but it finally dawned on me that the issue was a subsequent assertion failing causing the remote to be leaked. Looking at the config file on disk, the remote section is actually written correctly, but *somehow* the repo's view of it is not updated when the transaction completes, which causes the section count tests (in `with_fetchspec` and `with_empty_fetchspec`) to fail.

What's weird is that 1) this issue seems to be new on master, as the "old" version completes with 0 failures (at least locally on macOS) 2) this one's macOS build is green (Windows too), while my local build is not, with the same failing tests.

File: src/remote.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (pks-t) [reply src/remote.c]: Oh, yeah. That was my mistake, I somehow expected a `git_buf_detach` here. `git_buf_cstr` doesn't transfer memory ownership, so this is fine

**Comment by pks-t** [comment]
So yeah, the memory leak obviously isn't one, that was a thinko on my side. The only other two comments were only nice to have, so I'm approving. So as soon as CI's green and the conflict fixed I'd be happy to merge

**Comment by tiennou** [comment]
Rebased, and ready for review. I tore apart the transaction code I had introduced because somehow it's not a good idea to use those on our `config__weakptr` version. If you feel that's an interesting bug, I can whip up a test case for it.

File: src/remote.c [hunk 22]
```diff
@@ -189,58 +189,120 @@ static int canonicalize_url(git_buf *out, const char *in)
...
+{
+	GIT_INIT_STRUCTURE_FROM_TEMPLATE(
+		opts, version, git_remote_create_options, GIT_REMOTE_CREATE_OPTIONS_INIT);
+	return 0;
+}
+
+int git_remote_create_with_opts(git_remote **out, const char *url, const git_remote_create_options *opts)
+{
+	git_remote *remote = NULL;
+	git_transaction *txn = NULL;
```
Comment (pks-t) [inline src/remote.c]: This variable is now unused, which causes the CI to fail

Comment (tiennou) [reply src/remote.c]: Heh. Should be fixed now.

**Comment by pks-t** [comment]
Thanks a lot @tiennou for this awesome work!
